### PR TITLE
NH-93486: export traces via otlp

### PR DESCRIPTION
**Pull request by cleverchuk**
Enabling export of traces via `OTLP` by default unless user is pointing to `AO`. Added two new resource providers, `ApmResourceProvider` and `HostIdResourceProvider`. `ApmResourceProvider` is a refactor of existing code while `HostIdResourceProvider` turns the `HostId` data into a resource. All the created resources are eventually merged into a single resource.

This PR includes changes from #277 so that code can be reused. Please review #277 before this.

**Test Plan**:
Test services data [0](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1712644058766987264/overview?duration=21600), [1](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1712643928659124224/overview?duration=21600), [2](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1742334541200846848/logs?duration=21600) and [3](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1777406072376840192/overview?duration=21600)

### CODE REVIEW

File: custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java
Comment (cleverchuk) [inline custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java:1]: This resource is not being used. It looks like it causes issue with trace -> entity association.

**Comment by raphael-theriault-swi** [comment]
Could you rebase to make reviewing easier ?

File: custom/shared/src/main/java/com/solarwinds/opentelemetry/extensions/InboundMeasurementMetricsGenerator.java [hunk 1]
```diff
@@ -116,7 +118,13 @@ public void onEnding(ReadWriteSpan span) {
     final SpanContext parentSpanContext = spanData.getParentSpanContext();
     if (!parentSpanContext.isValid() || parentSpanContext.isRemote()) {
       span.setAttribute(TRANSACTION_NAME_KEY, TransactionNameManager.getTransactionName(spanData));
+      span.setAttribute(
+          AttributeKey.stringKey("TransactionName"),
+          TransactionNameManager.getTransactionName(spanData));
     }
+
+    span.setAttribute(
+        "Layer", String.format(LAYER_NAME_PLACEHOLDER, span.getKind(), span.getName().trim()));
```
Comment (raphael-theriault-swi) [inline custom/shared/src/main/java/com/solarwinds/opentelemetry/extensions/InboundMeasurementMetricsGenerator.java:127]: Very minor nitpick but most other attribute keys seem to be statics

File: custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java
Comment (raphael-theriault-swi) [reply custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java:1]: Should probably be used now that the ingestion issue was fixed ?

File: custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    setIfNotNull(builder::put, ResourceAttributes.HOST_NAME, hostId.getHostname());
+    setIfNotNull(
+        builder::put, ResourceAttributes.CLOUD_AVAILABILITY_ZONE, hostId.getEc2AvailabilityZone());
+    setIfNotNull(builder::put, ResourceAttributes.HOST_ID, hostId.getEc2InstanceId());
+    return Resource.create(builder.build());
+  }
+
+  private <V> void setIfNotNull(
+      BiConsumer<AttributeKey<V>, V> setter, AttributeKey<V> key, V value) {
```
Comment (raphael-theriault-swi) [inline custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java:145]: Another minor nitpick and more of a personal preference but given the setter is always the same in practice this feels a little more generic than it needs to be

File: custom/src/main/java/com/solarwinds/opentelemetry/extensions/initialize/ConfigurationLoader.java [hunk 3]
```diff
@@ -598,4 +637,9 @@ public static boolean shouldUseOtlpForMetrics() {
     return (enabled == null || enabled)
         && (collectorEndpoint == null || !collectorEndpoint.contains("appoptics.com"));
   }
+
+  public static boolean shouldUseOtlpForTraces() {
+    String collectorEndpoint = (String) ConfigManager.getConfig(ConfigProperty.AGENT_COLLECTOR);
+    return (collectorEndpoint == null || !collectorEndpoint.contains("appoptics.com"));
+  }
```
Comment (raphael-theriault-swi) [inline custom/src/main/java/com/solarwinds/opentelemetry/extensions/initialize/ConfigurationLoader.java:644]: Is this method ever used outside of tests ?

File: custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java
Comment (cleverchuk) [reply custom/src/main/java/com/solarwinds/opentelemetry/extensions/HostIdResourceProvider.java:1]: Will in a follow-on PR.

File: custom/src/main/java/com/solarwinds/opentelemetry/extensions/initialize/ConfigurationLoader.java (hunk 3 above)
Comment (cleverchuk) [reply custom/src/main/java/com/solarwinds/opentelemetry/extensions/initialize/ConfigurationLoader.java:644]: It doesn't look like it is.

File: custom/src/test/java/com/solarwinds/opentelemetry/extensions/initialize/ConfigurationLoaderTest.java [hunk 4]
```diff
@@ -390,16 +390,23 @@ void returnTrueWhenOtlpMetricExportIsNotDisabled() {
...
     assertFalse(ConfigurationLoader.shouldUseOtlpForMetrics());
     ConfigManager.removeConfig(ConfigProperty.AGENT_EXPORT_METRICS_ENABLED);
   }
+
+  @Test
+  void returnFalseWhenCollectorIsAOForTrace() throws InvalidConfigException {
+    ConfigManager.setConfig(ConfigProperty.AGENT_COLLECTOR, "collector.appoptics.com:443");
+    assertFalse(ConfigurationLoader.shouldUseOtlpForMetrics());
+    ConfigManager.removeConfig(ConfigProperty.AGENT_COLLECTOR);
+  }
```
Comment (raphael-theriault-swi) [inline custom/src/test/java/com/solarwinds/opentelemetry/extensions/initialize/ConfigurationLoaderTest.java:411]: This test has the same body as the one at lines 393-397

**Review by raphael-theriault-swi — CHANGES_REQUESTED** [review]

**Review by raphael-theriault-swi — APPROVED** [review]
LGTM

**Pull request merged**